In [1]:
import pandas as pd 
import seaborn as sns 
import numpy as np 
import io 


la validación cruzada es una técnica que nos permite obtener una evaluación más precisa y generalizada del rendimiento de un modelo de aprendizaje automático al entrenarlo y evaluarlo varias veces con diferentes divisiones del conjunto de datos. Es especialmente útil cuando se tiene un conjunto de datos limitado y se quiere tener una idea más precisa de cómo se comportará el modelo en datos no vistos, la validación cruzada reduce el sesgo y la variabilidad asociados con una única partición entre datos de entrenamiento y prueba, pero no puede garantizar completamente la independencia de los resultados.

In [3]:
df = pd.read_csv('/home/roy/Estadistica_inferencial/data/cancerdata_80367619-ee78-44cd-a7dc-7970bf56fa94.csv')
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0    ...        25.380          17.33           184.60      2019.0   
1    ...        24.990          23.41           158.80      1956.0   
2    ...        23.570          25.53           152.50      1709.0   
3    ...        14.910          26.50            98.87       567.7   
4    ...        22.540          16.67           152.20      1575.0   
..   ...           ...            ...              ...         ...   
564  ...        25.450          26.40           166.10      2027.0   
565  ...        23.690          38.25           155.00      1731.0   
566  ...        18.980          34.12           126.70      1124.0   
567  ...        25.740          39.42           184.60      1821.0   
568  ...         9.456          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564                0.2216       

In [13]:
X = df.iloc[:,2:] #variables numericas predictoras 
Y = df.iloc[:,1] # variable que queremos predecir 'categorica'


In [14]:
X.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   fractal_dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [15]:
Y.value_counts()
# como tenemos mas de 1 observacion podemos hacer este proceso del modelo de clasificacion 

diagnosis
B    357
M    212
Name: count, dtype: int64

In [24]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

# División de los datos en conjunto de entrenamiento y conjunto de prueba
# test_size=0.3 indica que el 30% de los datos se utilizará como conjunto de prueba,
# mientras que el 70% se utilizará como conjunto de entrenamiento.
# random_state=4 se utiliza para asegurar que se obtengan los mismos conjuntos de entrenamiento y prueba cada vez que se ejecute el código.
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=4)

# Creación de una instancia del clasificador del árbol de decisión
modelo = DecisionTreeClassifier()

# Entrenamiento del modelo utilizando el conjunto de entrenamiento
modelo.fit(X_train, Y_train)

# Evaluación del rendimiento del modelo utilizando el conjunto de prueba
# El método score() calcula la precisión del modelo en los datos de prueba
resultado = modelo.score(X_test, Y_test)

print(resultado)


0.9064327485380117


 el resultado obtenido muestra la precisión del modelo en el conjunto de prueba. En este caso, el modelo tuvo una precisión del 90.64%, lo que indica un buen rendimiento en términos de clasificación correcta de las muestras en el conjunto de prueba.

# validacion cruzada k fold

In [28]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

modelo = DecisionTreeClassifier()

# Creación de un objeto KFold para la validación cruzada con 10 folds (10 muestras)
KFold_validacion = KFold(10)

# Utilización de la validación cruzada para evaluar el rendimiento del modelo
resultados = cross_val_score(modelo, X, Y, cv=KFold_validacion)

# Imprimir los resultados de la validación cruzada para cada fold
print(resultados)
resultados.mean()


[0.92982456 0.9122807  0.89473684 0.94736842 0.92982456 0.96491228
 0.9122807  0.96491228 0.98245614 0.92857143]


0.93671679197995

los resultados muestran la precisión del modelo en cada fold de la validación cruzada, y el promedio de estas puntuaciones proporciona una medida general del rendimiento del modelo. En este caso, el modelo tuvo una precisión promedio del 93.67%, lo que indica que es un modelo con un buen rendimiento en general.